### DISCUSSION TITLE
How to add a custom request header？

### DISCUSSION THREAD

**Discussion by wangballball** [mitmproxy/mitmproxy#5155] [Q&A] [answered]
Here is my code.

      from mitmproxy import http
   
      class ModifyHeader:
   
            def request(self, flow: http.HTTPFlow):
                  flow.request.headers['test-request-header'] = 'test-request-header-value'
   
            def response(self, flow: http.HTTPFlow):
                  flow.response.headers['test-response-header'] = 'test-response-header-value'
   
   
        addons = [
            ModifyHeader()
        ]

mitmproxy==7.0.4

I want to add a custom request header, but it doesn't work. Strangely, the response header could be added successfully.
Who can help me? Thanks a lot.

**Comment by Prinzhorn** [comment, marked as answer]
> but it doesn't work

How do you verify this? Here's what `ncat` receives for me with your addon:

```
GET / HTTP/1.1
Host: 192.168.178.21:1337
User-Agent: curl/7.68.0
Accept: */*
Proxy-Connection: Keep-Alive
test-request-header: test-request-header-value


```

**Reply by wangballball** [reply]
I invoked google chrome via selenium, intercepted the request via mitmproxy, and then viewed the request and response via google chrome's developer mode. It can be seen that the response header has been modified successfully, but the request header has not been modified successfully. I guess this is because selenium has its own proxy, because I verified that I do not pass selenium but directly request my own service and then I can modify it successfully through mitmproxy

![image](https://user-images.githubusercontent.com/39551720/155913903-bedd4e2f-d34c-460b-b72d-488166c40797.png)
![image](https://user-images.githubusercontent.com/39551720/155913911-75467f17-1d70-4bbc-a316-2bbfe6e40151.png)

Thanks for your reply, it helped me a lot

**Reply by Prinzhorn** [reply]
> then viewed the request and response via google chrome's developer mode

the request you see in Chrome's network tab happened _before_ the request even reached mitmproxy

**Reply by wangballball** [reply]
So although it is not displayed on the network tab of Google Chrome, it is actually added successfully, right?
Thank you very much！

**Reply by Prinzhorn** [reply]
Correct

**Reply by wangballball** [reply]
Thank you very much and wish you a happy life🌹
